Adds the ability to jam cigarettes into the filter slot of a gas mask

Momo8289 opened the pull request:
## About The Pull Request
This PR makes it so that when a gas mask of any type (normal, explorer, mime, etc) has isn't holding any filters, and isn't covering the mouth, you can jam a cigarette in there and smoke it like normal. Cigs placed in a mask can be lit by using the lighting implement on the mask as you would a normal cig. The gas mask can't be adjusted while holding a cig (if the gas mask can be adjusted at all). Whether or not internals can be used depends on the mask and its adjusted state.
![image](https://github.com/tgstation/tgstation/assets/42454181/a1d00139-eef3-479a-b8f2-b12b531f6833)
A mime enjoying their favourite robusto cigars
## Why It's Good For The Game
First off, I just find the idea of replacing an air filter in a gas mask, something meant to protect you from harmful gasses and fumes, with a cancer-inducing burning stick of chemicals, to be really funny.

Secondly, and the reason that was the initial inspiration for this PR, is it allows mimes and clowns to smoke without removing their masks. This opens the door for mimes and clowns to use smoking as part of their character gimmick, without losing a pretty big part of their identity. 
## Changelog
:cl:
add: Cigarettes can now be placed in and smoked through gas masks, so long as the mask has no filters installed and is not covering the mouth.
/:cl:

Momo8289 committed:
sttuff

Momo8289 committed:
Merge branch 'tgstation:master' into cig-masks

Momo8289 committed:
Made visuals work

Momo8289 committed:
Can now actually smoke the cig

Momo8289 committed:
I don't even know but it all works now

Momo8289 committed:
Merge branch 'tgstation:master' into cig-masks

EricZilla commented:
Finally mimes can smoke

Helg2 commented:
Maybe also add possibility to put bubblegum in masks?

Momo8289 commented:
> Maybe also add possibility to put bubblegum in masks?

Implementing that in the same way doesn't really make a lot of sense. Shoving bubblegum into a gas mask with the intention to chew it wouldn't really work at all. The animation also wouldn't make much sense over a gas mask.

Helg2 commented:
> Shoving bubblegum into a gas mask with the intention to chew it wouldn't really work at all.

But you can eat through the mask.

dwasint commented:
unfathomably based

nukularpower commented:
I can finally be a clown with a stogie in his mouth??  this is the PR of the year imo

san7890 commented:
i've been waiting on this feature for a while now, thanks

san7890 commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +54,73 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+	return ..()
+
+/obj/item/clothing/mask/gas/adjustmask(mob/living/carbon/user)
+	if(cig)
+		user.visible_message(
+				span_warning("You need to remove [cig] before you can adjust the mask")
+		)
+		return
+	return ..()
```
```suggestion
	if(isnull(cig))
		return ..()
	user.visible_message(span_warning("You need to remove [cig] before you can adjust the mask!"))
```

This early return is a lot cleaner, as well as takes up less space. fixed up the formatting as well. could do a `balloon_alert()` instead too.

Momo8289 replied:
Gonna leave it as a span warning because that's what the code that was already there was using, if that makes sense?

Momo8289 replied:
Aaaand literally just saw your other comment now. I'll just switch them all to balloon alerts then.

san7890 commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +54,73 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
 	. = ..()
-	if(max_filters > 0)
+	if(cig)
+		. += "<span class='notice'>There is a [cig.name] jammed into the filter slot</span>"
```
```suggestion
		. += span_notice("There is a [cig.name] jammed into the filter slot.")
```

old pattern that never got fixed, how strange. This does the exact same thing. Also fixes punctuation.

san7890 commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +54,73 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+		return TRUE
+
+	if(cig)
+		var/c = cig.attackby(tool, user)
```
no new single-letter variables, please make it a bit more descriptive.

san7890 commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +54,73 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+		var/valid_wearer = istype(src.loc, /mob)
+
+		if(flags_cover & MASKCOVERSMOUTH)
+			user.visible_message(
+				span_warning("You can't do that while the mask's mouth is covered")
+			)
+			return ..()
+
+		if(max_filters <= 0 || cig)
+			user.visible_message(
+				span_warning("There's nowhere for that to fit")
+			)
+			return ..()
+		if(has_filter)
+			user.visible_message(
+				span_warning("You can't fit that in while there are filters in the mask")
+			)
```
these would all be better as balloon alerts. I'm also a big fan of multilining, but it doesn't make any sense when there's only one entry in the list, just looks uglier.

san7890 commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +54,73 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+
 /obj/item/clothing/mask/gas/attackby(obj/item/tool, mob/user)
+	if(istype(tool, /obj/item/clothing/mask/cigarette))
+		var/valid_wearer = istype(src.loc, /mob)
```
`ismob()` instead.

san7890 commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -64,7 +131,20 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+		user.put_in_hands(cig)
+		UnregisterSignal(cig, COMSIG_QDELETING)
+		cig = null
+		if(istype(src.loc, /mob))
```
`ismob()`

san7890 commented on code/game/objects/items/cigs_lighters.dm:
```diff
@@ -321,9 +321,38 @@ CIGARETTE PACKETS ARE IN FANCY.DM
-		reagents.remove_any(to_smoke)
-		return
+	// These checks are a bit messy but at least they're fairly readable
+	// Check if the smoker is a mob
+	if(!istype(smoker))
```
this actually does `!istype(smoker, /mob/living/carbon)` because you pathed it that way on line 323. Is that intentional? I think you should change 323 to be a typecast to `/mob` or `/mob/living` and then do something like `ismob()` or `isliving()`.

actually, line 323 is probably meant to be a carbon based on the whole accesses to `wear_mask`. I think this code needs some re-evaluation based on that.

Momo8289 replied:
323 is supposed to be a carbon, yeah. I'm guessing the confusion is from my comment. I'll update it

san7890 commented on code/game/objects/items/cigs_lighters.dm:
```diff
@@ -321,9 +321,38 @@ CIGARETTE PACKETS ARE IN FANCY.DM
+			reagents.remove_any(to_smoke)
+			return
+		// If it is, check if that mask is on a mob
+		if(!istype(smoker.loc, /mob/living/carbon))
```
this should be `iscarbon()`. it's also kinda confusing to keep checking the loc like this and making odd assertions, you should cache the loc as a local variable in this proc, like right after line 326 just do something like

`var/atom/potential_mask = smoker.loc`

that would just make everything hopefully make more sense, especially since we access `loc` more than once throughout this proc.

Momo8289 replied:
Will do

Momo8289 replied:
I think I've addressed this now, but let me know if it's still unclear.

san7890 reviewed (changes requested):
multiple nits. Some of it was a bit confusing for some reason, but after these reviews are addressed hopefully then I can figure out why some things look strange.

san7890 commented:
mark ready for review when this has all been addressed

Momo8289 committed:
Update code/modules/clothing/masks/gasmask.dm

Co-authored-by: san7890 <the@san7890.com>

Momo8289 committed:
Switched some text to use span_notice

Momo8289 committed:
switched over to balloon alerts

Momo8289 committed:
uses ismob

Momo8289 committed:
Fixes a syntax error

Momo8289 committed:
changed single letter var

Momo8289 committed:
Hopefully makes some code more clear

Momo8289 committed:
updated balloon alerts to be more concise

Momo8289 committed:
Merge branch 'master' into cig-masks

Momo8289 commented:
I've been told the failed check isn't my fault :+1:

san7890 commented:
sorry, i didn't accurately explain the balloon alert stuff. they must be complaint with the requirements in this document: https://hackmd.io/@tgstation/SJIle5wYd#balloon_alertballoon_alert_to_viewers

Momo8289 committed:
updated balloon text

Momo8289 commented:
> sorry, i didn't accurately explain the balloon alert stuff. they must be complaint with the requirements in this document: https://hackmd.io/@tgstation/SJIle5wYd#balloon_alertballoon_alert_to_viewers
> 
> this is to mean short, no capitalization nor punctuation, and informative. read the document and follow that for all of the things you've added.

Updated to fit the doc. It didn't say anything about them, so I left the apostrophes in, if that's okay, and added proper ending punctuation

MrMelbert commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -64,7 +123,20 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
 	has_filter = TRUE
 	return TRUE
 
+/obj/item/clothing/mask/gas/equipped(mob/M, slot)
+	. = ..()
+	if(cig)
+		cig.equipped(M, slot)
```
You have `/obj/item/clothing/mask/gas/equipped` implemented twice, which means you're duping calling `cig.equipped`

MrMelbert replied:
```suggestion
	cig?.equipped(M, slot)
```

MrMelbert commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +54,65 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+		. += span_notice("Currently there [LAZYLEN(gas_filters) == 1 ? "is" : "are"] [LAZYLEN(gas_filters)] filter\s with [get_filter_durability()]% durability.")
+		. += span_notice("The filters can be removed by right-clicking with an empty hand on [src].")
+
+/obj/item/clothing/mask/gas/proc/on_cig_destroyed()
+	SIGNAL_HANDLER
+	if(cig)
+		cig = null
+		if(istype(src.loc, /mob))
+			var/mob/wearer = src.loc
+			wearer.update_worn_mask()
```
It might be a bit easier to just implement gas mask `Exited` rather than register / unregister qdeleted signals

```dm
/obj/item/clothing/mask/gas/Exited(atom/movable/gone)
    . = ..()
    if(gone == cig)
        cig = null
        if(ismob(loc))
            var/mob/wearer = loc
            wearer.update_worn_mask()
```

MrMelbert commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +54,65 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+	SIGNAL_HANDLER
+	if(cig)
+		cig = null
+		if(istype(src.loc, /mob))
```
```suggestion
		if(ismob(loc))
```

MrMelbert commented on code/game/objects/items/cigs_lighters.dm:
```diff
@@ -321,9 +321,42 @@ CIGARETTE PACKETS ARE IN FANCY.DM
+
+		// If it is, check if that mob is actually wearing the mask
+		var/mob/living/carbon/mask_holder = potential_smoker
+		if(!mask_holder.wear_mask == smoker)
```
```suggestion
		if(mask_holder.wear_mask != smoker)
```

MrMelbert commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -26,10 +26,19 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+	if(!isinhands && cig)
+		. += cig.build_worn_icon(default_layer = FACEMASK_LAYER, default_icon_file = 'icons/mob/clothing/mask.dmi')
+	else
+		cut_overlays()
```
Do you know if this `cut_overlays` is necessary? Seems like it'd end up killing any other overlays added to the mask unnecessarily (since the cig overlay is simply not added if there is no cig)

MrMelbert commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +54,65 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
 
 /obj/item/clothing/mask/gas/attackby(obj/item/tool, mob/user)
+	if(istype(tool, /obj/item/clothing/mask/cigarette))
+		var/valid_wearer = ismob(src.loc)
```
```suggestion
		var/valid_wearer = ismob(loc)
```

MrMelbert commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +54,65 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+
+		cig = tool
+		if(valid_wearer)
+			cig.equipped(src.loc, ITEM_SLOT_MASK)
```
```suggestion
			cig.equipped(loc, ITEM_SLOT_MASK)
```

MrMelbert replied:
Note that this is misleading - just because the loc is a mob doesn't mean it's equipped to mask. it might be in hands. 

You may want to use `get_slot_by_item` and pass the slot it returns to equipped instead.

MrMelbert commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +54,65 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+			cig.equipped(src.loc, ITEM_SLOT_MASK)
+		RegisterSignal(cig, COMSIG_QDELETING, PROC_REF(on_cig_destroyed))
+		cig.forceMove(src)
+		var/mob/wearer = src.loc
```
```suggestion
		var/mob/wearer = loc
```

MrMelbert commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +54,65 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+
+	if(cig)
+		var/cig_attackby = cig.attackby(tool, user)
+		var/mob/wearer = src.loc
```
```suggestion
		var/mob/wearer = loc
```

MrMelbert replied:
Do you sanity check that loc is a mob before doing this chain?

MrMelbert commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -64,7 +123,20 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+		user.put_in_hands(cig)
+		UnregisterSignal(cig, COMSIG_QDELETING)
+		cig = null
+		if(ismob(src.loc))
+			var/mob/wearer = src.loc
```
```suggestion
		if(ismob(loc))
			var/mob/wearer = loc
```

Momo8289 committed:
removes unneeded cut_overlay

Momo8289 committed:
made not misleading

Momo8289 committed:
removed redundancy

Momo8289 committed:
changed misleading balloon alert

Momo8289 committed:
added sanity check

Momo8289 committed:
Merge branch 'master' into cig-masks

Momo8289 committed:
removed duplicate equipped

Momo8289 committed:
started using Exited

Momo8289 committed:
minor, mostly redundancy fixes

Momo8289 committed:
Merge branch 'master' into cig-masks

san7890 commented:
![image](https://github.com/tgstation/tgstation/assets/34697715/c4590b7c-57b7-4d3e-8f31-70f647c2ace5)

undraft when CI is passing

Momo8289 committed:
fixed indentation

Momo8289 committed:
Merge branch 'master' into cig-masks

MrMelbert commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -45,15 +52,65 @@ GLOBAL_LIST_INIT(clown_mask_options, list(
+
+	if(cig)
+		var/cig_attackby = cig.attackby(tool, user)
+		var/mob/wearer = loc
+		wearer.update_worn_mask()
```
I don't think you answered my other review, is `loc` guaranteed to be a mob if there is a cig in the mask? 
IE, does this runtime if you attackby the mask when it is unequipped-but-also-has-a-cig?

Momo8289 replied:
Ah, it does error. I'll fix that

Momo8289 committed:
fixes

weesnaw-ops commented:
I’ve been trying to do my chainsmoking clown gimmick for a while thank fuck

Momo8289 commented:
@MrMelbert Changes have been done for a while and it's going stale now. Could I get a merge?

Ghommie commented on code/game/objects/items/cigs_lighters.dm:
```diff
@@ -322,9 +322,42 @@ CIGARETTE PACKETS ARE IN FANCY.DM
+			reagents.remove_any(to_smoke)
+			return
+
+		var/atom/potential_smoker = smoker.loc
+
+		// If it is, check if that mask is on a carbon mob
+		if(!iscarbon(potential_smoker))
+			reagents.remove_any(to_smoke)
+			return
+
+		// If it is, check if that mob is actually wearing the mask
+		var/mob/living/carbon/mask_holder = potential_smoker
+		if(mask_holder.wear_mask != smoker)
+			reagents.remove_any(to_smoke)
+			return
+
+		// Update smoker to the wearer of the mask
+		smoker = potential_smoker
+
+	// Check if the smoker is actually wearing the cig
+	if(src != smoker.wear_mask)
+		// If they're not, check if the mask is a gas mask
+		if(istype(smoker.wear_mask, /obj/item/clothing/mask/gas))
+			// If it is, check if the cig is being held by that gas mask
+			var/obj/item/clothing/mask/gas/mask = smoker.wear_mask
+			if(src != mask.cig)
+				reagents.remove_any(to_smoke)
+				return
+		else
+			reagents.remove_any(to_smoke)
+			return
```
This is so confounding, can't something like this be done instead? (I suggest testing the changes locally if necessary cause I cannot deny being a bit bewildered) : 
```suggestion
		smoker = smoker.loc

		// If it is, check if that mask is on a carbon mob
		if(!istype(smoker) || smoker.get_item_by_slot(ITEM_SLOT_MASK) != loc)
			reagents.remove_any(to_smoke)
			return
```
The rest is mostly fine though.

Momo8289 replied:
Yeah that works, thanks

Momo8289 committed:
better checks

Momo8289 committed:
Merge branch 'master' into cig-masks
